subsystem/interlocks.py

```diff
@@ -1,60 +1,251 @@
-        ]
-        self.indicators = {
-            'active': tk.PhotoImage(file=resource_path("media/off_orange.png")),
-            'inactive': tk.PhotoImage(file=resource_path("media/on.png"))
-        }
+    def _create_interlocks_frame(self):
+        """Create the frame that will contain the interlock indicators"""
+        interlocks_frame = tk.Frame(self.interlocks_frame, highlightbackground="black")
+        interlocks_frame.grid(row=0, column=0, padx=0, pady=0, sticky="nsew")
+        
+        # Configure columns for indicator pairs (label + light)
+        num_columns = 22
+        for i in range(num_columns):
+            interlocks_frame.grid_columnconfigure(i, weight=1)
+        
+        return interlocks_frame
+
+    def _create_indicator_circle(self, frame, color):
+        """Create a circular indicator light"""
+        canvas = tk.Canvas(frame, width=30, height=30, highlightthickness=0)
+        canvas.grid(sticky='nsew')
+        oval_id = canvas.create_oval(5, 5, 25, 25, fill=color, outline="black")
+        return canvas, oval_id
+
+    def _create_indicators(self, frame):
+        """Create all indicator lights and their labels"""
+        for i, (name, _) in enumerate(self.INDICATORS.items()):
+            # Create label
+            tk.Label(frame, text=name, anchor="center").grid(
+                row=0, column=i*2, sticky='ew'
+            )
+            
+            # Create indicator light
+            canvas, oval_id = self._create_indicator_circle(frame, 'red')
+            canvas.grid(row=0, column=i*2+1, sticky='nsew')
+            self.INDICATORS[name] = (canvas, oval_id)
+
+    # updates indicator and logs updates
+    def update_interlock(self, name, safety, data):
+        """Update individual interlock indicator"""
+        if name not in self.INDICATORS or safety == None or data == None:
+            self.log("Invalid inputs to update_interlock", LogLevel.ERROR)
+
+        color = 'green' if (safety & data) == 1 else 'red'
+
+        if name in self.INDICATORS:
+            canvas, oval_id = self.INDICATORS[name]
+            current_color = canvas.itemcget(oval_id, 'fill')
+            if current_color != color:
+                canvas.itemconfig(oval_id, fill=color)
+                self.log(f"Interlock {name}: {current_color} -> {color}", LogLevel.INFO)
+
+    def _set_all_indicators(self, color):
+        """Set all indicators to specified color"""
+        if color == None or color == "":
+            self.log("Invalid inputs to _set_all_indicators", LogLevel.ERROR)
+
+        if self.INDICATORS:
+            for name in self.INDICATORS:
+                canvas, oval_id = self.INDICATORS[name]
+                current_color = canvas.itemcget(oval_id, 'fill')
+                if current_color != color:
+                    canvas.itemconfig(oval_id, fill=color)
+                    self.log(f"Interlock {name}: {current_color} -> {color}", LogLevel.INFO)
+
```

Comment:
In the update_data() method, shouldn't "if current_time - self.last_error_time > (self.update_interval / 1000):" check "if current_time - self.last_error_time > self.update_interval:" instead, given that current time and the timestamp of the last error are recorded in milliseconds as well..

Reply:
Hmm.. I think that this is OK the way it is currently.

[time.time()](https://docs.python.org/3/library/time.html#time.time) returns time in seconds (not ms), but `self.update_interval` is stored in milliseconds (see where it's initialized to 500 ms in the init method. So to compare these properly, we need to convert `update_interval` from milliseconds to seconds by dividing by 1000